theano/tensor/opt.py

```diff
@@ -5901,6 +5901,52 @@ def local_log_add(node):
                 return [ret]
 
 
+@register_stabilize
+@register_specialize
+@gof.local_optimizer([T.log])
+def local_log_sum_exp(node):
+    # log(sum_i(exp(x_i))) = x_max + log(sum_i(exp(x_i - x_max)))
+
+    if node.op != T.log:
+        return
+
+    sum_node = node.inputs[0].owner
+    # If the sum has keepdims=True, there might be a dimshuffle
+    if sum_node and isinstance(sum_node.op, T.DimShuffle):
```

Comment:
Yes I agree, there might be problems due to that. I guess the best solution would be to keep the DimShuffle op and re-apply it in the new graph.